src/bundle.c

```diff
@@ -2502,6 +2502,41 @@ gboolean load_manifest_from_bundle(RaucBundle *bundle, RaucManifest **manifest,
 	return res;
 }
 
+static gboolean read_complete_dm_device(gchar *dev, GError **error)
+{
+	int checkfd = -1;
+	char checkbuf[65536];
```

Comment:
Hmm, that's quite a large stack allocation.

Perhaps we should use O_DIRECT anyway, to avoid confusing the page cache algorithm by now accessing the data twice instead. It might also force out currently cached data. Take a look at `readable_sectors()` in `test/dm.c`. We should still use a large read size, than in the test (64kiB seems fine).

Reply:
I have taken the read code from `readable_sections()`.

`O_DIRECT` lead to a `Invalid argument` error while reading in the "plain" bundle case.

Reply:
Hmm, for `O_DIRECT`, you need to have 4kiB-alignment for the memory and the block to read. To you have the failing code somewhere we can take a look?

Otherwise, `--pre-check` should probably be skipped for plain bundles, as the signature verification verifies the full payload already (see `bundle->payload_verified`).

Reply:
Thinking about this again, perhaps we should skip O_DIRECT for now and see whether this is actually needed in practice. Nevertheless, you should use a buffer on the heap instead of 64kiB on the stack.

Reply:
Moved buffer to heap